src/backend/distributed/executor/multi_executor.c

```diff
@@ -96,6 +97,48 @@ CitusExecutorStart(QueryDesc *queryDesc, int eflags)
 }
 
 
+/*
+ * CitusExecutorRun is a wrapper around standard_ExecutorRun that keeps track of
+ * whether the current execution is in a function call (or SPI).
+ *
+ * When a distributed statement is executed in a transaction block, we need to
+ * make sure that we also run statements in a transaction block on the workers
+ * because they may have to be rolled back. The same applies to statements in
+ * a PL/pgSQL function, but there appears to be no Postgres function for
+ * detecting that we are in a PL/pgSQL function. However, since PL/pgSQL functions
+ * are executed through SPI, we can check the DestReceiver here.
+ */
+void
+CitusExecutorRun(QueryDesc *queryDesc, ScanDirection direction, uint64 count,
+				 bool executeOnce)
+{
+	DestReceiver *destReceiver = queryDesc->dest;
+	int originalLevel = FunctionCallLevel;
+
+	if (destReceiver->mydest == DestSPI)
```

Comment:
Is checking for `DestSPI` sufficient here? In particular, `DestSQLFunction` appears relevant (i.e. does this patch work with pure SQL functions (not PL/pgSQL?)…

I know @anarazel's comment on #659 is pretty old at this point, but at least the way I wrote it up sounded as though we wanted to check if `destReceiver->mydest >= DestSPI` (this would be bad code by our standards, I think, but you get the drift).

Is this patch meant to be comprehensive, or could there still be certain multi-statement transactions which do not do the right thing?

Reply:
I mostly had PL/pgSQL in mind while writing this patch, but including `DestSQLFunction` makes sense. The other tuple destinations don't indicate that there is a multi-statement transaction.

```
    DestSPI          - PL/pgSQL function and other languages
    DestTuplestore   - Mainly for things like RETURNING
    DestIntoRel      - CREATE TABLE tmp AS SELECT ...
    DestCopyOut      - COPY (SELECT ...) TO STDOUT
    DestSQLFunction  - SQL function output
    DestTransientRel - CREATE MATERIALIZED VIEW
    DestTupleQueue   - Parallel worker sends back to parent
```

The only other case I can think of is when there are multiple statements in a single query string. I'm not sure how to detect that reliably. Postgres does a `BeginImplicitTransactionBlock` to set the transaction state to `TBLOCK_IMPLICIT_INPROGRESS`, but there seems to be no way of getting that state. On the other hand, I expect that doing transactions in this manner is somewhat unusual so we can probably live with reduced support.

Reply:
Ah, `DestSQLFunction` is only set for the last statement in the function, not the first one :-/

Reply:
I think covering only PL/pgSQL for now is probably the way to go for now. It at least reduces the hazard quite a bit.